i#2006 generalize drcachesime: add two options for asynchronous trace…

zhaoqin opened the pull request:
Adds two options -num_threads and -queue_size to the tracer for asynchronous
trace write by the sideline threads.
This CL only adds the two options, and the real implementation is to be added.

derekbruening commented on clients/drcachesim/common/options.cpp:
```diff
@@ -42,6 +42,17 @@ droption_t<bool> op_offline
+ "0 means synchronized write without any sideline threads.");
+
+droption_t<unsigned int> op_queue_size
+(DROPTION_SCOPE_ALL, "queue_size", 4000, "The queue size for ",
```
"for "??

derekbruening replied:
Just SCOPE_CLIENT, right?

zhaoqin replied:
done

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/common/options.cpp:
```diff
@@ -42,6 +42,17 @@ droption_t<bool> op_offline
+droption_t<unsigned int> op_queue_size
+(DROPTION_SCOPE_ALL, "queue_size", 4000, "The queue size for ",
+ "For the offline analysis mode (when -offline is requested), specifies the size "
+ "of the queue for batching trace data to be written out by sideline threads.");
```
What are the units here?  If bytes this is way too small.  If buffers it seems large for a default: 128MB for 32K buffers.

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/common/options.cpp:
```diff
@@ -42,6 +42,17 @@ droption_t<bool> op_offline
+droption_t<bytesize_t> op_num_threads
+(DROPTION_SCOPE_ALL, "num_threads", 0, "Number of threads for writing traces",
+ "For the offline analysis mode (when -offline is requested), specifies the number "
+ "of sideline threads are used to write traces file out.  "
```
grammar: s/are/to be/
grammar: s/traces file/trace files/

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/common/options.cpp:
```diff
@@ -42,6 +42,17 @@ droption_t<bool> op_offline
  "for later offline analysis.  No simulator is executed.");
 
+droption_t<bytesize_t> op_num_threads
+(DROPTION_SCOPE_ALL, "num_threads", 0, "Number of threads for writing traces",
```
Shouldn't this be SCOPE_CLIENT?

zhaoqin replied:
done

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -2462,7 +2462,7 @@ if (CLIENT_INTERFACE)
         torunonly_ci(tool.drcacheoff.${testname} ${exetgt} drcachesim
           "offline-${testname}.c" # for templatex basename
-          "-offline" "" "")
+          "-offline -num_threads 0" "" "")
```
Isn't this the default?

zhaoqin replied:
was trying to test if the option can be specified.
removed.

zhaoqin commented:
PTAL

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
+static volatile uint queue_tail = 0;
+static void *sideline_exit_event;
+struct buf_entry_t {
+    file_t  file;  // file descriptor, 0 means the end of profiling.
```
nit: INVALID_FILE, not 0

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
+/* sideline */
+static drvector_t buf_queue;
+static uint queue_size = 0;
+static uint queue_size_mask = 0;
```
Hmm, no, if this is based on queue_size, then queue_size must be the length of the vector?  Why do we need a separate var for that: it's already in buf_queue, right?

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
+static uint queue_size_mask = 0;
+static volatile uint queue_front = 0;
+static volatile uint queue_tail = 0;
+static void *sideline_exit_event;
```
Don't we need a separate one per thread?

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
+        }
+        drvector_lock(&buf_queue);
+        if (((queue_front + 1) & queue_size_mask) != queue_tail) {
+            drvector_set_entry(&buf_queue, queue_front, entry);
```
So the invariant is that the front is empty?  Maybe a comment on that up above?  How about an assert that the current front file is INVALID_FILE?

derekbruening replied:
Was this done?

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
+        if (entry->data != NULL) {
+            file_ops_func.write_file(entry->file, entry->data, entry->size);
+            dr_raw_mem_free(entry->data, max_buf_size);
+        } else if (entry->file != 0) {
```
INVALID_FILE

derekbruening replied:
This one?

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
+             */
+            file_ops_func.close_file(entry->file);
+        } else {
+            profiling = false;
```
So at process exit you put a single file==0 entry in the queue -- but only one thread will grab that, and you want to exit all threads.  This doesn't seem like it will scale to more than one thread easily as a method of exiting.  This is one reason why being incremental by having just one thread may not be the best approach: we need to be sure the design works with a real pool.  IMHO it is not worth putting in this sentinel entry scheme if it will not work with multiple threads.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -272,7 +354,9 @@ write_trace_data(void *drcontext, byte *towrite_start, byte *towrite_end)
-        if (file_ops_func.handoff_buf != NULL) {
+        if (op_num_threads.get_value()) {
+            trace_buf_enqueue(data->file, towrite_start, size);
+        } else if (file_ops_func.handoff_buf != NULL) {
```
So you've put a priority here: the handoff is going to be ignored if there's a thread pool.  That should be reflected in its docs.

Maybe we should just remove the handoff feature completely, since it's not needed anymore: it can be duplicated by using a thread pool and having file_ops_func.write do whatever you want.

I also think the thread pool should be the default.

derekbruening commented on clients/drcachesim/common/options.cpp:
```diff
@@ -42,6 +42,20 @@ droption_t<bool> op_offline
+ "For the offline analysis mode (when -offline is requested), specifies the number "
+ "of sideline threads to be used to write trace files out.  "
+ "Only 0 and 1 is currently supported.  "
+ "0 means synchronized write without any sideline threads.  ");
```
IMHO the pool should be on by default, for code simplicity.

I'm not sure about only supporting 1 thread: we need to know that the design here supports many threads, and seeing the stop-profiling scheme below I'm not sure that it does.  Adding >1 thread seems like just a few lines of code: array of events, for loop when touching those events.  It seems better to include multi-thread support now, to ensure the design works (again see below where I'm not sure it does).

zhaoqin replied:
A threading pool with more than one thread has a complex problem, i.e, sideline thread vs open files.
The simplest way is one open file per sideline thread, which requires us to change raw2trace.
If we want to keep current one file per app-thread, writing to same file with multiple threads is complex.
I will go with the one file per sideline thread.

derekbruening replied:
I don't remember seeing any mention of this in the CL?  Did I miss it?  E.g., the scheme for putting an entry in the queue to terminate the sideline thread function isn't marked as temporary, so the reader can't tell why only one thread is supported and why we can't just add an array and for loops to easily increase the pool to 100 threads.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -872,14 +963,27 @@ event_exit(void)
+    if (op_offline.get_value()) {
+        if (op_num_threads.get_value() > 0) {
+            trace_buf_enqueue((file_t)0, NULL, 0);  // end of profiles
+            dr_event_wait(sideline_exit_event);
```
Needs to be a for loop, one event per thread, right?

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -951,6 +1055,21 @@ init_offline_dir(void)
+    queue_size = i;
+    if (queue_size < 2) {
+        NOTIFY(0, "Usage error: queue size too small\n");
+        DR_ASSERT(false);
```
dr_abort

derekbruening replied:
There are other comments like this that I did not see any reply to.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -1044,6 +1163,19 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
+        get_queue_size();
+        drvector_init(&buf_queue, queue_size, false, NULL);
+        dr_create_client_thread(sideline_run, NULL);
+        sideline_exit_event = dr_event_create();
```
for loop around these two lines

zhaoqin commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
 
+/* sideline */
+static drvector_t buf_queue;
+static uint queue_size = 0;
```
done

zhaoqin commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
+static uint queue_size = 0;
+static uint queue_size_mask = 0;
+static volatile uint queue_front = 0;
+static volatile uint queue_tail = 0;
```
done

zhaoqin commented:
Just updated the threading pool implementation.
PTAL.

zhaoqin commented:
I put comment for queue_front/queue_tail (replaced with
queue_put/queue_get), which I feel clearer.

On Mon, Apr 24, 2017 at 9:30 PM, Derek Bruening <notifications@github.com>
wrote:

> *@derekbruening* commented on this pull request.
> ------------------------------
>
> In clients/drcachesim/tracer/tracer.cpp
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113091277>:
>
> > +trace_buf_enqueue(file_t file, void *data, ssize_t size)
> +{
> +    bool done = false;
> +    buf_entry_t *entry = (buf_entry_t *)dr_global_alloc(sizeof(*entry));
> +    entry->file = file;
> +    entry->data = data;
> +    entry->size = size;
> +    do {
> +        /* quick racy check without locking */
> +        while (((queue_front + 1) & queue_size_mask) == queue_tail) {
> +            /* the queue is full, waiting */;
> +            dr_thread_yield();
> +        }
> +        drvector_lock(&buf_queue);
> +        if (((queue_front + 1) & queue_size_mask) != queue_tail) {
> +            drvector_set_entry(&buf_queue, queue_front, entry);
>
> Was this done?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113091277>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkcDPeoJc9n0--B16LYEo6jU2371lx6ks5rzUy6gaJpZM4Mu6lp>
> .
>

zhaoqin commented:
I kind of rewriting a lot, including removing entry->file.

On Mon, Apr 24, 2017 at 9:30 PM, Derek Bruening <notifications@github.com>
wrote:

> *@derekbruening* commented on this pull request.
> ------------------------------
>
> In clients/drcachesim/tracer/tracer.cpp
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113091284>:
>
> > +{
> +    buf_entry_t *entry;
> +    bool profiling = true;
> +    do {
> +        /* quick racy check without locking */
> +        while (queue_tail == queue_front) {
> +            /* the queue is empty, waiting */
> +            dr_thread_yield();
> +        }
> +        entry = trace_buf_dequeue();
> +        if (entry == NULL)
> +            continue;
> +        if (entry->data != NULL) {
> +            file_ops_func.write_file(entry->file, entry->data, entry->size);
> +            dr_raw_mem_free(entry->data, max_buf_size);
> +        } else if (entry->file != 0) {
>
> This one?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113091284>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkcDEldw-0AzHlp_15RdV0_fdw_1Qcoks5rzUzBgaJpZM4Mu6lp>
> .
>

zhaoqin commented on suite/tests/CMakeLists.txt:
```diff
@@ -2459,47 +2459,58 @@ if (CLIENT_INTERFACE)
-        set(tool.drcacheoff.${testname}_basedir
+      macro (torunonly_drcacheoff testname exetgt sideline)
+        if (${sideline})
+          set(tname "${testname}-sideline")
```
removed since the default is the sideline.
Will add test without sideline

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -211,48 +217,187 @@ static uint     tls_offs;
+    do {
+        /* quick racy check without locking */
+        while (queue_get == queue_put) {
+            /* the queue is empty, waiting */
```
Similarly to the queue being full, it would be better to have a proper kernel-involved wait here.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -211,48 +217,187 @@ static uint     tls_offs;
+            }
+            dr_raw_mem_free(entry.data, max_buf_size);
+        } else {
+            DR_ASSERT(entry.size == 0);
```
A comment would help to understand what this is.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -211,48 +217,187 @@ static uint     tls_offs;
+    } while (profiling);
+
+    dr_mutex_lock(mutex);
+    dr_mutex_unlock(mutex);
```
?  lock, unlock?

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -262,49 +407,41 @@ atomic_pipe_write(void *drcontext, byte *pipe_start, byte *pipe_end)
 write_trace_data(void *drcontext, byte *towrite_start, byte *towrite_end)
 {
+    if (op_num_threads.get_value() > 0) {
+        circular_buf_enqueue(towrite_start, towrite_end - towrite_start,
```
See other comment about moving this

zhaoqin replied:
rename it to process_trace_data

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -786,55 +934,21 @@ event_thread_init(void *drcontext)
-        /* put buf_base to TLS plus header slots as starting buf_ptr */
-        BUF_PTR(data->seg_base) = data->buf_base + buf_hdr_slots_size;
+        byte *buf = (byte *)
+            dr_raw_mem_alloc(max_buf_size, DR_MEMPROT_READ | DR_MEMPROT_WRITE, NULL);
```
bug: this is leaked.  The prior code seems superior, using a static buffer.

zhaoqin replied:
Cannot use static buffer, because this buffer maybe enqueued, which will be freed on dequeue.
It is a leak, and fixed.

derekbruening replied:
Can it be smaller?  This uses 64K or whatnot when it only really needs a tiny amount.

zhaoqin replied:
Only if we change the code for the buffer free part and pass the size around.
IMHO the code is much simpler, also we will free it very soon after dequeue it from the queue.
It should add no peak memory usage which is bounded by the queue size.

What if there are thousands threads during attach?
It may slowdown the attach, because the thread init event will be blocked by the enqueue.

derekbruening replied:
> What if there are thousands threads during attach?
> It may slowdown the attach, because the thread init event will be blocked by the enqueue.

This is a separate topic, right?  This is about the thread "pool" having just one thread for online and worries about it having worse performance than no pool at all?

derekbruening replied:
Did you measure the performance for online and for offline?  We have some numbers for SPEC in the tracker and it would be good to see what effect the pool has on a simple app like that.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -854,35 +968,57 @@ event_thread_exit(void *drcontext)
+    NOTIFY(1, "drmemtrace exiting process " PIDFMT"; traced " SZFMT" bytes.\n",
+           dr_get_process_id(), num_bytes);
+
+
```
nit: extra blank line

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -854,35 +968,57 @@ event_thread_exit(void *drcontext)
+        if (op_offline.get_value())
+            close_thread_file(sideline_args[i].file);
+    }
+    if (op_num_threads.get_value() > 0) {
```
suggest: move for loop inside this if

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -951,13 +1087,119 @@ init_offline_dir(void)
+        /*do nothing*/;
+    queue_size = i;
+    if (queue_size < 2) {
+        NOTIFY(1, "Usage error: queue size too small\n");
```
0 not 1

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -951,13 +1087,119 @@ init_offline_dir(void)
+        dr_global_alloc(op_num_threads.get_value() * sizeof(sideline_exit_events[0]));
+    for (i = 0; i < op_num_threads.get_value(); i++) {
+        sideline_exit_events[i] = dr_event_create();
+        sideline_args[i].id   = i;
```
nit: extra spaces

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -951,13 +1087,119 @@ init_offline_dir(void)
+        sideline_args[i].id   = i;
+        if (op_offline.get_value())
+            sideline_args[i].file = create_thread_file((ptr_int_t)i);
+        dr_create_client_thread(sideline_run, &sideline_args[i]);
```
Check the return value

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -971,11 +1213,18 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
+        }
+        if (op_num_threads.get_value() > 1) {
+            NOTIFY(0, "Usage error: only one sideline thread is supported "
+                   "for online simulation\n");
```
Why?  Please explain in a comment: is this temporary or permanent?
Is this documented in options.cpp?

zhaoqin replied:
done.
only one named pipe, so one thread makes sense, right?

derekbruening replied:
Will something not work correctly if there's more than one thread in the pool?  Or it will work, you just think there's no perf improvement?

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -211,48 +217,187 @@ static uint     tls_offs;
+            entry->data = data;
+            entry->size = size;
+            entry->tid = tid;
+            drvector_set_entry(&circular_buf_queue, queue_put, entry);
```
Isn't this redundant?

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -951,13 +1087,119 @@ init_offline_dir(void)
+    queue_get = 0;
+}
+
+void event_fork_init(void *drcontext)
```
style violation: \n

Also, should be static

zhaoqin replied:
done

zhaoqin commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -211,48 +217,187 @@ static uint     tls_offs;
-        /* Avoid future buffer output. */
-        op_max_trace_size.set_value(data->bytes_written - 1);
-        return;
+        dr_abort();
```
done

zhaoqin commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -318,59 +455,25 @@ memtrace(void *drcontext, bool skip_size_cap)
         }
-        if (op_offline.get_value()) {
+        if (op_num_threads.get_value() > 0) {
             write_trace_data(drcontext, pipe_start, buf_ptr);
```
done

zhaoqin commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -762,16 +865,61 @@ event_pre_syscall(void *drcontext, int sysnum)
+    size  = instru->append_thread_file_header((byte *)buf);
+
+    DR_ASSERT(size > 0 && size < MAXIMUM_PATH);
+    file_ops_func.write_file(file, (void *)buf, (size_t)size);
```
done

zhaoqin commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -1038,11 +1268,24 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
     redzone_size = instru->sizeof_entry() * MAX_NUM_ENTRIES;
     max_buf_size = trace_buf_size + redzone_size;
-    buf_hdr_slots_size = instru->sizeof_entry() * BUF_HDR_SLOTS;
+    /* get the header size by faking an unit header write */
```
done

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -2440,7 +2440,7 @@ if (CLIENT_INTERFACE)
       torunonly_ci(tool.reuse ${ci_shared_app} drcachesim
         "reuse_distance.c" # for templatex basename
-        "-ipc_name drtestpipe6 -simulator_type reuse_distance -reuse_distance_threshold 256" "" "")
+        "-ipc_name drtestpipe6 -num_threads 0 -simulator_type reuse_distance -reuse_distance_threshold 256" "" "")
```
OK but it is confusing b/c this looks like there's some relationship to the reuse-distance tool: like the reuse-distance tool doesn't support the thread pool or sthg.  I would suggest either a comment explaining, or adding a new test run that looks just an existing run except it has -num_threads 0.

zhaoqin replied:
done, add a separate test instead.

zhaoqin commented:
It probably will work but may lose accuracy, because more than one threads
are writing to the pipe at the same time.
I could remove the check

On Wed, Apr 26, 2017 at 12:28 AM, Derek Bruening <notifications@github.com>
wrote:

> *@derekbruening* commented on this pull request.
> ------------------------------
>
> In clients/drcachesim/tracer/tracer.cpp
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113363234>:
>
> > @@ -971,11 +1213,18 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
>                 droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
>          dr_abort();
>      }
> -    if (!op_offline.get_value() && op_ipc_name.get_value().empty()) {
> -        NOTIFY(0, "Usage error: ipc name is required\nUsage:\n%s",
> -               droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
> -        dr_abort();
> -    } else if (op_offline.get_value() && op_outdir.get_value().empty()) {
> +    if (!op_offline.get_value()) {
> +        if (op_ipc_name.get_value().empty()) {
> +            NOTIFY(0, "Usage error: ipc name is required\nUsage:\n%s",
> +                   droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
> +            dr_abort();
> +        }
> +        if (op_num_threads.get_value() > 1) {
> +            NOTIFY(0, "Usage error: only one sideline thread is supported "
> +                   "for online simulation\n");
>
> Will something not work correctly if there's more than one thread in the
> pool? Or it will work, you just think there's no perf improvement?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113363234>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkcDP-WKX2T3rlcNp-tERh3Sxx7PIvCks5rzsfcgaJpZM4Mu6lp>
> .
>

zhaoqin commented:
No, still working on the corrrectness, saw some hangs.

On Wed, Apr 26, 2017 at 11:04 AM, Derek Bruening <notifications@github.com>
wrote:

> *@derekbruening* commented on this pull request.
> ------------------------------
>
> In clients/drcachesim/tracer/tracer.cpp
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113477635>:
>
> >      } else {
>          /* pass pid and tid to the simulator to register current thread */
> -        proc_info = (byte *)buf;
> -        DR_ASSERT(BUFFER_SIZE_BYTES(buf) >= 3*instru->sizeof_entry());
> -        proc_info += instru->append_thread_header(proc_info, dr_get_thread_id(drcontext));
> -        proc_info += instru->append_tid(proc_info, dr_get_thread_id(drcontext));
> -        proc_info += instru->append_pid(proc_info, dr_get_process_id());
> -        write_trace_data(drcontext, (byte *)buf, proc_info);
> -
> -        /* put buf_base to TLS plus header slots as starting buf_ptr */
> -        BUF_PTR(data->seg_base) = data->buf_base + buf_hdr_slots_size;
> +        byte *buf = (byte *)
> +            dr_raw_mem_alloc(max_buf_size, DR_MEMPROT_READ | DR_MEMPROT_WRITE, NULL);
>
> Did you measure the performance for online and for offline? We have some
> numbers for SPEC in the tracker and it would be good to see what effect the
> pool has on a simple app like that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DynamoRIO/dynamorio/pull/2319#discussion_r113477635>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkcDJ3KEU1q4IxHVxjTwOkSKuO4ktqHks5rz1zjgaJpZM4Mu6lp>
> .
>

zhaoqin commented:
PTAL
there are some non-trivial changes, so it would be better to take another round of review.

derekbruening commented on clients/drcachesim/common/options.cpp:
```diff
@@ -36,12 +36,32 @@
  "for later offline analysis.  No simulator is executed.");
 
+droption_t<unsigned int> op_num_threads
+(DROPTION_SCOPE_CLIENT, "num_threads", IF_WINDOWS_ELSE(0, 1),
```
This needs a comment: why is it different for Windows?

Also, if the pool is a perf hit (esp w/ only 1 thread) for offline (but we expect to make it a net win with tuning, right?), should it be disabled by default on all platforms to avoid slowing everyone down?

If this really depends on online vs offline then have the default set in C++ code and describe it in the docs here.

zhaoqin replied:
add comment

zhaoqin commented:
New commit message:
i#2006 generalize drcachesime: add threading pool support for writing out traces

Added threading pool for asynchronous trace writing out.
- added two options for threading pool configuration,
- used a circular buffer for producer-consumer queue,
- added fork handling,
- updated trace format to write trace per sideline thread instead of app thread,
- updated raw2trace to handle the new trace format,
- added new tests for different threading pool configurations,
- removed handoff callbacks and corresponding tests.

derekbruening commented on clients/drcachesim/tracer/raw2trace.cpp:
```diff
@@ -403,13 +403,13 @@ raw2trace_t::merge_and_process_thread_files()
     // The current thread we're processing is tidx.  If it's set to thread_files.size()
```
current file

I guess leaving as "tidx" is fine: less code churn.

derekbruening commented on clients/drcachesim/tracer/raw2trace.cpp:
```diff
@@ -431,56 +431,46 @@ raw2trace_t::merge_and_process_thread_files()
                 }
             }
-            VPRINT(2, "Next thread in timestamp order is %u @0x" ZHEX64_FORMAT_STRING
-                   "\n", (uint)tids[next_tidx], times[next_tidx]);
```
Why was this removed?  I found it useful for seeing the operation: which file is next.

derekbruening commented on clients/drcachesim/tracer/raw2trace.cpp:
```diff
@@ -431,56 +431,46 @@ raw2trace_t::merge_and_process_thread_files()
-                WARN("Input file for thread %d is truncated", (uint)tids[tidx]);
-                in_entry.extended.type = OFFLINE_TYPE_EXTENDED;
-                in_entry.extended.ext = OFFLINE_EXT_TYPE_FOOTER;
+                VPRINT(2, "Finish read file %d\n", tidx);
```
Finished reading

derekbruening commented on clients/drcachesim/tracer/raw2trace.cpp:
```diff
@@ -431,56 +431,46 @@ raw2trace_t::merge_and_process_thread_files()
-                offline_entry_t entry;
-                if (thread_files[tidx]->read((char*)&entry, sizeof(entry)) ||
-                    !thread_files[tidx]->eof())
-                    FATAL_ERROR("Footer is not the final entry");
```
We're now missing two checks:

1) That each thread ends with a footer: this is important to know whether we had thread truncation.  This was checked above at lines 458-462.  I think we have to put this check back in (will have to iterate the threads at the end I guess), along with the warning and workaround.

2) That each file is not truncated.  We probably need a new kind of footer now, a file footer entry.

zhaoqin replied:
done

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -211,16 +219,185 @@ static uint     tls_offs;
+    buf_entry_t *entry;
+    do {
+        /* quick racy check without locking */
+        while (((queue_put + 1) & queue_size_mask) == queue_get) {
```
You said you saw major performance issues involving contention on the queue: please add comments about that so we know this code needs improvement.

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -2380,23 +2400,25 @@ if (CLIENT_INTERFACE)
           "${PROJECT_SOURCE_DIR}/clients/drcachesim/tests")
         set(tool.drcachesim.threads_rawtemp ON) # no preprocessor
-        set(tool.drcachesim.threads_timeout 150) # This test is long.
+        set(tool.drcachesim.threads_timeout 200) # This test is long.
```
The pool slows all of these down so much we need longer timeouts??  That's not good.  Sounds like this should definitely be off by default until its performance is improved.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -211,16 +219,185 @@ static uint     tls_offs;
+write_trace_data(file_t file, void *data, ssize_t size)
+{
+    ssize_t written = file_ops_func.write_file(file, data, size);
+    DR_ASSERT(written == size);
```
This used to be NOTIFY0 with an understandable message and dr_abort, which seems better than "written==size" being printed

derekbruening commented:
Also, the commit message should talk about the performance, since that's a big TODO item: reading the current commit message it sounds like this thread pool is all finished.

zhaoqin commented:
PTAL

derekbruening commented on clients/drcachesim/tracer/raw2trace.cpp:
```diff
@@ -451,21 +456,33 @@ raw2trace_t::merge_and_process_thread_files()
+                VPRINT(2, "Finish reading file %d at pos %u\n",
+                       tidx, (int)thread_files[tidx]->tellg());
+                // We need do another failed read to make ->eof() return true.
+                if (thread_files[tidx]->read((char*)&in_entry, 1)) {
```
The previous code that you deleted seems better, "if (read entry || !eof) fatal..."

derekbruening commented on clients/drcachesim/tracer/raw2trace.cpp:
```diff
@@ -510,6 +528,13 @@ raw2trace_t::merge_and_process_thread_files()
         }
     } while (file_count > 0);
+    // If we used threading pool during profiling, the trace for a thread may be
+    // seen after the thread exit entry, so we can only check whether the unique
```
What?  But you can't put a thread exit entry in the final trace here and then put a memref entry for that thread *afterward*.  Is that what's happening due to the timestamp changes?  We have to enforce that order, don't we?

derekbruening commented on clients/drcachesim/tracer/raw2trace.cpp:
```diff
@@ -510,6 +528,13 @@ raw2trace_t::merge_and_process_thread_files()
+    // tid and tid exit matches.
+    VPRINT(1, "%u thread(s) are seen\n", (uint)tid_set.size());
+    if (tid_set != tid_exit_set) {
+        WARN("Profile might be corrupted.");
```
A more specific message, please!  Again the prior code was IMHO better: "Trace for thread %d is truncated".

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -247,7 +247,10 @@ circular_buf_enqueue(void *data, ssize_t size, thread_id_t tid)
         /* quick racy check without locking */
         while (((queue_put + 1) & queue_size_mask) == queue_get) {
-            /* XXX: use futex/semaphore for better performance. */
+            /* XXX: there is a big slowdown for the online simulation
```
I thought you said offline was slow too?  You plan to put the numbers in the issue tracker, right?

fhahn commented:
Can one of the admins verify this patch?

fhahn commented:
Can one of the admins verify this patch?
